blockchain/blockindex_test.go

```diff
@@ -310,3 +311,200 @@ func TestAncestorSkipList(t *testing.T) {
 		}
 	}
 }
+
+// TestWorkSorterCompare ensures the work sorter less and hash comparison
+// functions work as intended including multiple keys.
+func TestWorkSorterCompare(t *testing.T) {
+	lowerHash := mustParseHash("000000000000c41019872ff7db8fd2e9bfa05f42d3f8fee8e895e8c1e5b8dcba")
+	higherHash := mustParseHash("000000000000d41019872ff7db8fd2e9bfa05f42d3f8fee8e895e8c1e5b8dcba")
+	tests := []struct {
+		name     string     // test description
+		nodeA    *blockNode // first node to compare
+		nodeB    *blockNode // second node to compare
+		wantCmp  int        // expected result of the hash comparison
+		wantLess bool       // expected result of the less comparison
+
+	}{{
+		name: "exactly equal, both data",
+		nodeA: &blockNode{
+			hash:            *mustParseHash("0000000000000000000000000000000000000000000000000000000000000000"),
+			workSum:         big.NewInt(2),
+			status:          statusDataStored,
+			receivedOrderID: 0,
+		},
+		nodeB: &blockNode{
+			hash:            *mustParseHash("0000000000000000000000000000000000000000000000000000000000000000"),
+			workSum:         big.NewInt(2),
+			status:          statusDataStored,
+			receivedOrderID: 0,
+		},
+		wantCmp:  0,
+		wantLess: false,
+	}, {
+		name: "exactly equal, no data",
+		nodeA: &blockNode{
+			hash:            *mustParseHash("0000000000000000000000000000000000000000000000000000000000000000"),
+			workSum:         big.NewInt(2),
+			receivedOrderID: 0,
+		},
+		nodeB: &blockNode{
+			hash:            *mustParseHash("0000000000000000000000000000000000000000000000000000000000000000"),
+			workSum:         big.NewInt(2),
+			receivedOrderID: 0,
+		},
+		wantCmp:  0,
+		wantLess: false,
+	}, {
+		name: "a has more cumulative work, data, same order, higher hash",
+		nodeA: &blockNode{
+			hash:            *higherHash,
+			workSum:         big.NewInt(4),
+			status:          statusDataStored,
+			receivedOrderID: 0,
+		},
+		nodeB: &blockNode{
+			hash:            *lowerHash,
+			workSum:         big.NewInt(2),
+			receivedOrderID: 0,
+		},
+		wantCmp:  1,
+		wantLess: false,
+	}, {
+		name: "a has less cumulative work, same order, lower hash, b has data,",
+		nodeA: &blockNode{
+			hash:            *lowerHash,
+			workSum:         big.NewInt(2),
+			receivedOrderID: 0,
+		},
+		nodeB: &blockNode{
+			hash:            *higherHash,
+			workSum:         big.NewInt(4),
+			status:          statusDataStored,
+			receivedOrderID: 0,
+		},
+		wantCmp:  -1,
+		wantLess: true,
+	}, {
+		name: "a has same cumulative work, data, same order, lower hash",
+		nodeA: &blockNode{
+			hash:            *lowerHash,
+			workSum:         big.NewInt(2),
+			status:          statusDataStored,
+			receivedOrderID: 0,
+		},
+		nodeB: &blockNode{
+			hash:            *higherHash,
+			workSum:         big.NewInt(2),
+			receivedOrderID: 0,
```

Comment:
`name` doesn't correspond to actual test case (claims nodeA and nodeB have the same data but only nodeA has `statusDataStored`

Reply:
I made the description clearer for this an a few others.  It was meant to "a has same cumulative work", "a has data", "a has same order", etc, but I see that wasn't clear.